Changes to a CCX course should emit a course_published event

cguardia opened the pull request:
Custom Courses for EdX should eventually behave and appear to other parts of the system as regular courses. Sending a course published event when a CCX course is changed is a step in that direction.

@ormsbee recommended a way to do this, and this PR attempts to implement that. This change was discussed in more detail in https://github.com/jazkarta/edx-platform/issues/80.

The change only affects CCXs and should be transparent for everything else.

edx-webhook commented:
Thanks for the pull request, @cguardia! I've created [OSPR-725](https://openedx.atlassian.net/browse/OSPR-725) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

sarina commented on lms/djangoapps/ccx/tests/test_tasks.py:
```diff
@@ -0,0 +1,58 @@
+"""
+test tasks
```
Seems like a better docstring could be used here

sarina commented:
@ormsbee is there someone else who can review this PR besides you? Your plate is overfull with OSPR reviews right now.

ormsbee commented on lms/djangoapps/ccx/tasks.py:
```diff
@@ -0,0 +1,43 @@
+    Consume signals that indicate course published. If course already a CCX, do nothing.
+    """
+    if not isinstance(course_key, CCXLocator):
+        send_ccx_course_published.delay(course_key)
```
We've had problems before because the celery message serialization is configured to use JSON and the keys won't work properly. Please send the string serialization, or take some other measure to make sure it works right in JSON.

ormsbee commented:
@sarina: Anyone from Platform can take this one. It's a fairly straightforward change -- the big issue is just making sure that we test it right, since celery is configured a bit differently between devstack and our prod environments.

ormsbee commented:
It would be good, for instance, to add test cases to make sure that course structures were actually generated for this based on the publish signal. Similarly with CourseOverviews, which are also publish-triggered.

cguardia commented:
Added the suggested tests for course_overview and course_structure handling. Also addressed the possible json issue mentioned by sending the unicode key instead of the course id object.

sarina commented:
@cguardia looks like a CCX unit test is still failing

cguardia commented:
@sarina fixed failing test, sorry for falling to a case of "it worked on my computer".

cguardia commented:
@ormsbee I think this is ready, any other suggestions before the last rebase?

sarina commented:
@cguardia just FYI Dave is on vacation for the next two weeks. I've sent the ticket to platform team and they will groom/review when they are able to.

cguardia commented:
@sarina thanks a lot.

cpennington commented on lms/djangoapps/ccx/tests/test_tasks.py:
```diff
@@ -0,0 +1,102 @@
+    def call_fut(self, course_key):
+        """Call the function under test
+        """
+        from ..tasks import send_ccx_course_published
```
Nit: why the local import?

cguardia replied:
uh, none that I can recall at the moment...I'll try to get it out of the way.

cpennington commented on lms/djangoapps/ccx/tests/test_tasks.py:
```diff
@@ -0,0 +1,102 @@
+        """
+        with mock_signal_receiver(SignalHandler.course_published) as receiver:
+            self.call_fut(self.course.id)
+            self.assertEqual(receiver.call_count, 1)
```
Is it worth validating that this receiver sends out signals to *all* ccxs (rather than just one)?

cguardia replied:
It probably won't hurt.

cpennington commented:
I had a couple questions, but this otherwise looks good.

cguardia commented:
jenkins run lettuce

cguardia commented:
jenkins run bokchoy

sarina commented:
@cpennington can you take another look at this, and merge or suggest next steps?
